### PR TITLE
Community - Option to set beneficiaries on post

### PR DESCRIPTION
**Pull request by eonwarped**
References Issue #2076 .

Adds the ability to specify beneficiaries on posts.

What it does:
* Adds UI controls to add beneficiaries and assign percentages to posts on creation.
* Autocomplete on following users.
* Validations-- no duplicates, cannot set self to beneficiary, percentages > 0, total <= 100

What's missing: 
* Tooltip explaining what this does.
* Language Translations for error messages.
* Tests? Suggestions for what tests should be added?
* Showing % payout going to beneficiaries after payout (though it currently doesn't do it anyway for posts that have it through other means)

How it looks:
beneficiaryinfoadded

### CODE REVIEW

**Comment by eonwarped** [comment]
Ran blackbox test locally and it passes.

docker run --network container:condenser steemit/condenser-blackboxtest:latest
_XSERVTransmkdir: ERROR: euid != 0,directory /tmp/.X11-unix will not be created.
yarn run v1.5.1
$ nightwatch --config nightwatch.json
Starting selenium server... started - PID:  34

[Homepage] Test Suite
=========================

Running:  Ensure homepage loads without strange errors
 ✔ Testing if element <a.Header__login-link> contains text: "Login".
 ✔ Passed [ok]: error log is reasonable

OK. 2 assertions passed. (9.102s)

[Ssr] Test Suite
====================
Running:  Ensure all of the standard routes return HTTP 200
 ✔ Passed [equal]: Response for /privacy.html is 200
 ✔ Passed [equal]: Response for /faq.html is 200
 ✔ Passed [equal]: Response for /recover_account_step_1 is 200
 ✔ Passed [equal]: Response for /support.html is 200
 ✔ Passed [equal]: Response for /tos.html is 200
 ✔ Passed [equal]: Response for /recover_account_step_2 is 200
 ✔ Passed [equal]: Response for /market is 200
 ✔ Passed [equal]: Response for /promoted is 200
 ✔ Passed [equal]: Response for /about.html is 200
 ✔ Passed [equal]: Response for /login.html is 200
 ✔ Passed [equal]: Response for /create_account is 200
 ✔ Passed [equal]: Response for /created is 200
 ✔ Passed [equal]: Response for /approval is 200
 ✔ Passed [equal]: Response for /hot is 200
 ✔ Passed [equal]: Response for /submit.html is 200
 ✔ Passed [equal]: Response for /change_password is 200
 ✔ Passed [equal]: Response for /~witnesses is 200
 ✔ Passed [equal]: Response for /waiting_list.html is 200

OK. 18 assertions passed. (2.905s)

OK. 20  total assertions passed. (12.102s)
 ✔ Passed [equal]: Response for / is 200
 ✔ Passed [equal]: Response for /trending is 200
Done in 14.39s.

**Comment by pkattera** [comment]
Thanks for this contribution @eonwarped 

Adding beneficiaries to a post is a useful feature for advanced users. However, every configuration option we add to the posting editor adds a burden to most users (especially new users). Adding two advanced features together (rewards type + beneficiaries) feels complex and requires improvements to the overall UX. 

We'd need to implement a solution that allows most people to skip past, while still providing advanced users access. For example, on the Post editor screen, we could replace the posting reward options with a link called 'Advanced options'. This would allow new users to skip past it. 

screen shot 2018-04-20 at 10 15 53 am

Clicking 'Advanced options' would trigger a lightbox modal with the posting rewards options.

advanced-options

Like your implementation, users could add more beneficiaries.

advanced-options2

Ideally, we should auto-complete account names to prevent errors (like we do on the transfers screen)

advanced-options3

----

Acceptance criteria:
- Improvements to UX as outlined above
- The 'beneficiaries' feature should be implemented as a new React component, which could be dropped into the new modal.

Obviously, this increases the scope of your initial PR. Let me know if you're still interested in helping out with this?

**Comment by eonwarped** [comment]
Cool, I like this. I can start looking into it, though I'm still learning
as I go :). I think it shouldn't be too hard borrowing from similar
structures, but for sure it will not be a quick matter.

**Comment by eonwarped** [comment]
@pkattera Do you prefer a click-wall styling (as referenced in #2512) or a modal as designed here?

Also is there a newer user autocomplete module that's done, or does that have to be built still?

**Comment by pkattera** [comment]
Once there are multiple items (rewards + beneficiaries), using a modal to group together the advanced settings works better. 

I don't believe there is a newer auto-complete module, so it would be best to reference the one in the wallet 'transfer' screen. Like in Transfers, I'd be fine if we don't include the user's profile image.

screen shot 2018-05-01 at 12 35 16 am

**Comment by roadscape** [comment]
Hey @eonwarped is this something you're still working on?

**Comment by eonwarped** [comment]
I haven't started yet but was thinking about setting aside time later
today. There's another issue tracking that will use the modal right?

**Comment by roadscape** [comment]
@eonwarped I don't believe there's a separate issue, @pkattera has posted the designs here

**Comment by eonwarped** [comment]
Oh... now I see. We removed everything that would have gone in advanced
options :)
I suppose that means there would only be beneficiaries in this modal I'm
working on (and possibly the ability to override setting defaults).

**Comment by eonwarped** [comment]
Here's what I have so far:

* New link to pop up the modal.
advancedsettings

* Adding an account.
addedaccount

* How it looks when adding many, up to 8.

addlotsofbeneficiaries

* After saving it with 1 beneficiary:

indicatoronreply

-----

Some notes:

* If we want to put this in comments as well, we'd need to store these in state keyed by form ID (which I believe shouldn't be too hard). 
* Where I stored the state will most likely want to be moved (user.current.reply.beneficiaries?)
* Styling seems off from the mock, is there a general movement for applying styling to these h2,h3,h4? Unsure where to do so.
* On editing posts, I do not show beneficiaries in read-only mode. That seems useful but will leave out of this set of changes.
* Looking into how to test the setting of it.

**Comment by eonwarped** [comment]
(Small updates-- changing the location of the state. Also fixing a bug where the state is not cleared on submission)

**Comment by roadscape** [comment]
Hey @eonwarped, looking good!

> If we want to put this in comments as well, we'd need to store these in state keyed by form ID (which I believe shouldn't be too hard). Where I stored the state will most likely want to be moved (user.current.reply.beneficiaries?)

It's probably low-priority for comments, but we do save draft state in localStorage keyed by author/permlink. For new top-level posts the key is `replyEditorData-submitStory`, which I suppose is the only case we really need to handle right now. I'd consider beneficiaries part of draft state so we should keep them together; it makes the ability to save multiple drafts easy in the future.

> Styling seems off from the mock, is there a general movement for applying styling to these h2,h3,h4? Unsure where to do so.

Generally we should just aim to keep styling markup similar to other modals, and we'll polish the CSS later.

> On editing posts, I do not show beneficiaries in read-only mode. That seems useful but will leave out of this set of changes.

Seems fine to me!

> I noticed that the way that the 50/50 comment options works is a little awkward and may need to be moved into this Advanced Setting modal that's initially populated by the default and then swapped out. For now, I left it as is.

The rewards option was moved to the Settings page and is now saved server-side. We'll need to figure something out a way to add it back at some point.. probably to the Advanced Settings modal, see https://github.com/steemit/condenser/issues/2779 for details.

**Comment by eonwarped** [comment]
> I'd consider beneficiaries part of draft state so we should keep them together; it makes the ability to save multiple drafts easy in the future.
Yup, I made sure that it works, and I did end up moving it to a place keyed by the formId.

> The rewards option was moved to the Settings page and is now saved server-side. We'll need to figure something out a way to add it back at some point.. probably to the Advanced Settings modal, see #2779 for details.
Sounds good. I can definitely work to move it after this.

Still trying to figure out how best to test some of this functionality as well.

**Comment by eonwarped** [comment]
I'd say this is ready for review now.

**Comment by roadscape** [comment]
Thank you @eonwarped excellent progress, will perform an in-depth review ASAP.

File: src/app/components/cards/BeneficiarySelector.test.jsx [hunk 1]
```diff
@@ -0,0 +1,162 @@
+import React from 'react';
+import { translate } from 'app/Translator';
```
[the remark this answers is not in the document]
Comment (eonwarped) [reply src/app/components/cards/BeneficiarySelector.test.jsx]: without some import from app/Translator, the validation tests will fail with not being able to look up translations from counterpart/tt. I suppose 'translate' itself isn't being used though, let me see if I can adjust it.

**Comment by roadscape** [comment]
Hey @eonwarped, I don't see contact info on your profile. Could you please email me at tim@steemit.com? Would like to discuss potential work opportunities.

**Comment by eonwarped** [comment]
Reached out by email, mentioning this random string *djGkJwk1$?!

**Comment by eonwarped** [comment]
Hey all, in parallel I worked on adding in payout type as well, would it make sense for me to just throw it in here as well?

I also got it to work with both comments and posts. But it may be worth first introducing the payout type by itself first without beneficiaries. What do you think?

**Comment by roadscape** [comment]
@eonwarped if I understand correctly, you are suggesting to introduce the payout switch before beneficiaries?

Also, it looks like this implementation saves beneficiaries to react state; I'm wondering if we should keep it in localStorage with the rest of the draft state. The concern here is that if the user refreshes the page, their post remains intact but they might not be aware the beneficiaries were reset.

**Comment by roadscape** [comment]
Let's get the minimum set of features merged in and get a fresh PR for further additions. If you're suggesting payout types should go in first, it would require re-organizing this PR. What do you think?

**Comment by eonwarped** [comment]
I'm okay with adding this first, and adding payout switch later (I was suggesting payouts first in case beneficiaries was too complicated to add first, I can send a parallel PR with just the payout change if we want to do that).

About the react state vs local storage, I did store it in localstorage in this implementation as well, and on load it sets the react state again.  It also puts an indicator for when it thinks beneficiaries are set  "(x beneficiary(ies) set".

I think I can edit the text to say "Beneficiaries: # set."  instead, which I had in my further changes, to be similar to how "Payout: 100% Power Up." looks.

**Comment by eonwarped** [comment]
@roadscape  Since I suspect it will be a lot simpler to review payout first, I pulled it out here:  #2879. We can probably keep this PR for beneficiary issue specifically, and I can update this once the other one is reviewed.

**Comment by eonwarped** [comment]
Now up to date with advanced settings. Diff is not so clean, but current state from advanced settings can be seen with
https://github.com/steemit/condenser/pull/2700/files/3ebb8c68ff67632be55d8a1330636c161cea73c0..8073e9de2cc3548845d9c1cd92b4796ee5545701

**Comment by rebuildman** [comment]
I followed the changes made here
(https://github.com/steemit/condenser/pull/2700/files/3ebb8c68ff67632be55d8a1330636c161cea73c0..8073e9de2cc3548845d9c1cd92b4796ee5545701) and I updated my condenser, stayed like this. Perfect.

congifnew
confgavancadas
only error I found when we will comment on some post, I get this error, Transaction broadcast error.
RPCError: Assert Exception:beneficiaries.size(): Must specify at least one beneficiary 

errorr
Maybe I jumped something I should do?

**Comment by eonwarped** [comment]
@rebuildman

Ack. That's a good catch!

Line 1034 in ReplyEditor.jsx should check for length as well. I will update when I get a chance.

**Comment by rebuildman** [comment]
@eonwarped  when the correction is made, let me know, I really liked this option that we can make available to the user, thank you, you're doing a great job.

**Comment by eonwarped** [comment]
@rebuildman tested locally, resolved now. Thanks for finding the problem!

**Comment by rebuildman** [comment]
Good morning, @eonwarped , I believe I found one more problem, I did 2 tests onten posting, I set up the beneficiary chooses 1 and 2 beneficiary, and I went to steemd to check if everything was okay, but did not show the beneficiary configured.

**Comment by eonwarped** [comment]
@rebuildman any steps to reproduce? I set two beneficiaries just now and it worked as expected.

**Comment by rebuildman** [comment]
@eonwarped in my condenser I just made the changes quoted here  " https://github.com/steemit/condenser/pull/2700/files/3ebb8c68ff67632be55d8a1330636c161cea73c0..8073e9de2cc3548845d9c1cd92b4796ee5545701 " will it be that I need to make changes to all the files found here " https://github.com/steemit/condenser/pull/2700/files "?

**Comment by eonwarped** [comment]
@rebuildman can you push your changes into your GitHub repository? You can also do a diff with the changes that are in my branch.

**Comment by eonwarped** [comment]
Hasn't been merging recent changes as of late. If we want to adopt I can go
ahead and bring it up to date.

But the last status awhile ago was that it was ready to review.

**Comment by roadscape** [comment]
Ok cool, let's see if @pkattera wants to move forward with this.

**Comment by roadscape** [comment]
Local branch for testing: `eonwarped-addbeneficiary`

**Comment by roadscape** [comment]
Works great. Local PR: #3517

**Pull request closed without merging**